java-tests/validating/src/test/kotlin/io/spine/test/options/when/ProtoTimestampWhenSpec.kt

```diff
@@ -0,0 +1,242 @@
+                    value.addAll(severalFutureTimes)
+                }
+            }
+
+            @Test
+            fun `pass, if not restricted at all`() = assertValidationPasses {
+                anyProtoTimestamps {
+                    value.addAll(severalFutureTimes)
+                }
+            }
+        }
+
+        @Nested
+        inner class `with a single past stamp within the future stamps` {
+
+            private val severalFutureAndPast = listOf(futureTime(), pastTime(), futureTime())
+
+            @Test
+            fun `throw, if restricted to be in future`() = assertValidationFails {
+                futureProtoTimestamps {
+                    value.addAll(severalFutureAndPast)
+                }
+            }
+
+            @Test
+            fun `throw, if restricted to be in past`() = assertValidationFails {
+                pastProtoTimestamps {
+                    value.addAll(severalFutureAndPast)
+                }
+            }
+
+            @Test
+            fun `pass, if not restricted at all`() = assertValidationPasses {
+                anyProtoTimestamps {
+                    value.addAll(severalFutureAndPast)
+                }
+            }
+        }
+
+        @Nested
+        inner class `with a single future stamp within the past stamps` {
+
+            private val severalPastAndFuture = listOf(pastTime(), futureTime(), pastTime())
+
+            @Test
+            fun `throw, if restricted to be in future`() = assertValidationFails {
+                futureProtoTimestamps {
+                    value.addAll(severalPastAndFuture)
+                }
+            }
+
+            @Test
+            fun `throw, if restricted to be in past`() = assertValidationFails {
+                pastProtoTimestamps {
+                    value.addAll(severalPastAndFuture)
+                }
+            }
+
+            @Test
+            fun `pass, if not restricted at all`() = assertValidationPasses {
+                anyProtoTimestamps {
+                    value.addAll(severalPastAndFuture)
+                }
+            }
+        }
+    }
+}
+
+private fun pastTime(): Timestamp {
+    val current = Timestamps.now()
+    val past = Timestamps.subtract(current, FIFTY_MILLIS)
+    return past
+}
+
+private fun futureTime(): Timestamp {
+    val current = Timestamps.now()
+    val future = Timestamps.add(current, FIFTY_MILLIS)
+    return future
+}
+
+/**
+ * Protobuf [Duration] of fifty milliseconds.
+ *
+ * To shift the time into the past or future, we add or subtract a difference of this amount.
+ *
+ * There are two reasons for choosing fifty milliseconds:
+ *
+ * 1. The generated code uses `io.spine.base.Time.currentTime()` to get the current timestamp
+ * for comparison. In turn, this method relies on `io.spine.base.Time.SystemTimeProvider`
+ * by default, which has millisecond precision.
+ * 2. Adding too small amount of time to make the stamp denote "future" might be unreliable.
```

Comment:
So why don't we choose two days instead?

Reply:
To be closer to the precision threshold. `two days` is already a seconds-based precision. Any value less than 1s would confirm millisecond precision during comparison.

Reply:
I've bumped to 500ms in case of slow VMs on CI.